lib/spack/spack/cmd/logs.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import gzip
+import os
+import shutil
+import sys
+
+import spack.cmd
+import spack.util.compression as compression
+from spack.cmd.common import arguments
+
+description = "print out logs for packages"
+section = "basic"
+level = "long"
+
+
+def setup_parser(subparser):
+    arguments.add_common_arguments(subparser, ["spec"])
+
+
+def _dump_byte_stream_to_stdout(instream):
+    if sys.platform == "win32":
+        # https://docs.python.org/3/library/sys.html#sys.stdout unconditionally
+        # recommends using .buffer to write binary data to sys.stdout, but
+        # https://docs.python.org/3/library/io.html#io.TextIOBase.buffer claims
+        # it is not available in all implementations, so only try accessing it
+        # on Windows, where stdout.fileno() is not useful
+        try:
+            outstream = sys.stdout.buffer
+        except AttributeError:
+            raise spack.main.SpackCommandError(
+                "This command must be invoked in a"
+                " context where stdout is accessible"
+                f" as a binary stream (stdout = {type(sys.stdout)})"
+            )
+    else:
+        outstream = os.fdopen(sys.stdout.fileno(), "wb", closefd=False)
+
+    shutil.copyfileobj(instream, outstream)
+
+
+def dump_build_log(package):
+    with open(package.log_path, "rb") as f:
+        _dump_byte_stream_to_stdout(f)
+
+
+def _logs(cmdline_spec, concrete_spec):
+    if concrete_spec.installed:
+        log_path = concrete_spec.package.install_log_path
+    elif os.path.exists(concrete_spec.package.stage.path):
+        dump_build_log(concrete_spec.package)
+        return
+    else:
+        raise spack.main.SpackCommandError(f"{cmdline_spec} is not installed or staged")
+
+    if not os.path.exists(log_path):
+        raise spack.main.SpackCommandError(f"No logs are available for {cmdline_spec}")
```

Comment:
Nitpick: `open(..., "rb")` is going to happen anyways, `os.path.exists` does not imply that `open` works, so the way people usually write these things is to just

```python
try:
  with open(log_path) as f:
    ...
except OSError as e:
  raise SpackCommandError(f"No logs available for ...: {e}") from e
```

Reply:
EDIT: I restored the `if` because it is useful for generating sensible error messages when using `compression.extension_from_file` (which I argue is reasonable here in https://github.com/spack/spack/pull/42202#discussion_r1470093550)

Reply:
```
if spec.installed
```
that checks the database ;)